generated/generated.go

```diff
@@ -15,67 +14,57 @@
 //
-// 	The text must appear as the first line of a properly formatted Go // comment,
-// 	and that comment must appear before but not be attached to the package clause
-// 	and before any /* */ comment. This is similar to the rules for build tags.
-//
-// 	The comment line must match the case-sensitive regular expression (in Go syntax):
+// 	Generated files are marked by a line of text that matches
+// 	the regular expression, in Go syntax:
 //
 // 		`^// Code generated .* DO NOT EDIT\.$`
 //
 // 	The .* means the tool can put whatever folderol it wants in there,
 // 	but the comment must be a single line and must start with `Code generated`
 // 	and end with `DO NOT EDIT.`, with a period.
 //
-// If the source couldn't be read, the error indicates the specific
-// failure. If the source was read but syntax errors were found,
-// the result is estimated on a best effort basis from a partial AST.
+// 	The text may appear anywhere in the file.
 //
-// TODO: Decide on best policy of what to do in case of syntax errors
-// being encountered during parsing.
+// If the source couldn't be read, the error indicates the specific
+// failure.
 func ParseFile(filename string) (hasGeneratedComment bool, err error) {
-	fset := token.NewFileSet()
-	f, err := parser.ParseFile(fset, filename, nil, parser.PackageClauseOnly|parser.ParseComments)
-	if f == nil { // Can only happen when err != nil.
+	f, err := os.Open(filename)
+	if err != nil {
 		return false, err
 	}
-Outer:
-	for _, cg := range f.Comments {
-		if cg == f.Doc {
-			// If we've reached the package comment, don't look any further,
-			// because the generated comment must be before that.
-			break
+	defer f.Close()
+	br := bufio.NewReader(f)
+	for {
+		s, err := br.ReadBytes('\n')
+		if err == io.EOF {
+			return containsComment(s), nil
+		} else if err != nil {
+			return false, err
 		}
-		// Check if this comment group is a match.
-		// The text must appear as the first line of a properly formatted line comment (//-style).
-		if len(cg.List[0].Text) >= smallestMatchingComment &&
-			strings.HasPrefix(cg.List[0].Text, "// Code generated ") &&
-			strings.HasSuffix(cg.List[0].Text, " DO NOT EDIT.") &&
-			fset.Position(cg.List[0].Pos()).Column == 1 {
-
+		s = s[:len(s)-1] // Trim newline.
+		if containsComment(s) {
 			return true, nil
 		}
-		// Ensure none of the comments in this comment group are general comments (/*-style).
-		for _, c := range cg.List {
-			if strings.HasPrefix(c.Text, "/*") {
-				// If we've reached a general comment (/*-style), don't look any further,
-				// because the generated comment must be before that.
-				break Outer
-			}
-		}
 	}
-	return false, nil
+}
+
+// containsComment reports whether a line of Go source code s (without newline character)
+// contains the generated comment.
+func containsComment(s []byte) bool {
+	return len(s) >= smallestMatchingComment &&
+		bytes.HasPrefix(s, []byte("// Code generated ")) &&
```

Comment:
https://play.golang.org/p/SRAr9N8JRi

There's no heap allocation either way, which is why the alloc numbers didn't change. There's just the `stringtoslicebyte` runtime call still there, which is copying the string memory into a temporary array on the stack. That only happens if the string is 32 bytes or fewer:

```
        case OARRAYRUNESTR:
                a := nodnil()

                if n.Esc == EscNone {
                        // Create temporary buffer for string on stack.
                        t := typArray(Types[TUINT8], tmpstringbufsize)

                        a = nod(OADDR, temp(t), nil)
                }

                n = mkcall("slicerunetostring", n.Type, init, a, n.Left)

                // stringtoslicebyte(*32[byte], string) []byte;
        case OSTRARRAYBYTE:
                a := nodnil()

                if n.Esc == EscNone {
                        // Create temporary buffer for slice on stack.
                        t := typArray(Types[TUINT8], tmpstringbufsize)

                        a = nod(OADDR, temp(t), nil)
                }

                n = mkcall("stringtoslicebyte", n.Type, init, a, conv(n.Left, Types[TSTRING]))
```

```
$ git grep stringtoslicebyte
src/cmd/compile/internal/gc/builtin.go: {"stringtoslicebyte", funcTag, 44},
src/cmd/compile/internal/gc/builtin/runtime.go:func stringtoslicebyte(*[32]byte, string) []byte
src/cmd/compile/internal/gc/walk.go:            // stringtoslicebyte(*32[byte], string) []byte;
src/cmd/compile/internal/gc/walk.go:            n = mkcall("stringtoslicebyte", n.Type, init, a, conv(n.Left, Types[TSTRING]))
```

So you're getting lucky because your strings to match are short, as they often are in reality, which is why the compiler's optimization often works.